packages/liveblocks-yjs/src/index.ts

```diff
@@ -0,0 +1,253 @@
+import type {
+  BaseUserMeta,
+  Json,
+  JsonObject,
+  LsonObject,
+  Room,
+} from "@liveblocks/client";
+import { Base64 } from "js-base64";
+import { Observable } from "lib0/observable";
+import * as Y from "yjs";
+
+type RoomEvent = {
+  type: "REFRESH";
+};
+
+type LiveblocksYjsOptions = {
+  httpEndpoint?: string;
+};
+
+type RefreshResponse = {
+  updates: string[];
+  lastUpdate: number;
+};
+
+type MetaClientState = {
+  clock: number;
+  lastUpdated: number;
+};
+
+/**
+ * This class will store YJS awareness in Liveblock's presence under the __yjs key
+ * IMPORTANT: The yjs awareness protocol uses ydoc.clientId to reference users
+ * to their respective documents. To avoid mapping yjs clientIds to liveblock's connectionId,
+ * we simply set the clientId of the doc to the connectionId. Then no further mapping is required
+ */
+export class Awareness extends Observable<any> {
+  private room: Room<JsonObject, LsonObject, BaseUserMeta, Json>;
+  public doc: Y.Doc;
+  public clientID: number;
+  public states: Map<number, any> = new Map();
+  // Meta is used to keep track and timeout users who disconnect. Liveblocks provides this for us, so we don't need to
+  // manage it here. Unfortunately, it's expected to exist by various integrations, so it's an empty map.
+  public meta: Map<number, MetaClientState> = new Map();
+  // _checkInterval this would hold a timer to remove users, but Liveblock's presence already handles this
+  // unfortunately it's expected to exist by various integrations.
+  public _checkInterval: number = 0;
+
+  private othersUnsub: () => void;
+  constructor(
+    doc: Y.Doc,
+    room: Room<JsonObject, LsonObject, BaseUserMeta, Json>
+  ) {
+    super();
+    this.doc = doc;
+    this.room = room;
+    this.clientID = doc.clientID;
+    this.othersUnsub = this.room.events.others.subscribe(({ event }) => {
+      // When others are changed, we emit an event that contains arrays added/updated/removed.
+      if (event.type === "leave") {
+        // REMOVED
+        this.emit("change", [
+          { added: [], updated: [], removed: [event.user.connectionId] },
+          "local",
+        ]);
+      }
+
+      if (event.type === "enter") {
+        // ADDED
+        this.emit("change", [
+          { added: [event.user.connectionId], updated: [], removed: [] },
+          "local",
+        ]);
+      }
+
+      if (event.type === "update") {
+        // UPDATED
+        this.emit("change", [
+          { added: [], updated: [event.user.connectionId], removed: [] },
+          "local",
+        ]);
+      }
+    });
+  }
+
+  destroy(): void {
+    this.emit("destroy", [this]);
+    this.othersUnsub();
+    this.setLocalState(null);
+    super.destroy();
+  }
+
+  getLocalState(): JsonObject | null {
+    const presence = this.room.getPresence();
+    if (Object.keys(this.room.getPresence()).length === 0) {
+      return null;
+    }
+    return presence["__yjs"] as JsonObject | null;
```

Comment:
If there is presence, but not a `__yjs` field in there, could this potentially return `undefined`, even though TypeScript thinks it can only be `JsonObject | null`?

Reply:
good catch! thanks!